fix use of default tech database flags

Goober5000 opened the pull request:
The "default" tech database flags indicate whether entries are in the tech room by default, that is, when the table is originally parsed and the campaign is started.  Flags may be changed later, so that when the tech database is "reset to default" the flags can be returned to the known state as specified in the tables.

A bug was introduced back in commit c924e3997a (!) based on a misinterpretation of what "default" meant.  This fixes the bug and clarifies the flags.

Fixes #4390.

notimaginative commented:
I don't remember the exact details, but I believe that change was actually intended to fix a bug related to something with pilot files (mod switching or related??). I do agree that the fix itself is a bug though. I wish I could remember what issue it was supposed to fix so we'd know to look for it if it pops up again. There's probably a comment thread buried somewhere in the HLP forums about it.

Goober5000 commented:
It could have been related to the fact that the pilot file needs to store a tri-state but the tech db flag is a boolean.  The pilot file code has been rewritten since then, of course, but this new code looks relevant:

```
	// write whether it differs from the default for ships
	for (auto &si : Ship_info) {
		if (si.flags[Ship::Info_Flags::In_tech_database] == si.flags[Ship::Info_Flags::Default_in_tech_database]) {
			state = TechroomState::DEFAULT;
		} else if (si.flags[Ship::Info_Flags::In_tech_database]) {
			state = TechroomState::ADDED;
		} else {
			state = TechroomState::REMOVED;
		}

		cfwrite_ubyte((ubyte) state, cfp);
	}
```

notimaginative commented:
Quite possible. The initial "fix" was done after the first big pilot file rewrite and most likely isn't something that applies to the current code. I did try to address a lot of the strange issues that popped up before/after the first pilot file rewrite while initially working on the newer code, though I'm not sure how much of that ended up being in the final code.

But this LGTM.
